File: cases/11-heavy-reporting-blocks-operations/go/app/main.go
// Caso 11 — Reportes pesados que bloquean operacion — stack Go 1.23.
//
// Legacy: el reporte corre sin acotar en la goroutine del request. Con varios
// reportes concurrentes se ocupan todos los procesadores logicos y /order-write
// se degrada.
// Isolated: el reporte pasa por un limitador de concurrencia; como maximo N
// reportes corren a la vez y siempre queda CPU para la operacion.
//
// El contraste que este stack aporta, y por que este caso NO se traduce literal:
//
//   Java y .NET aislan con **pools de threads separados**: un ThreadPoolExecutor
//   de 4 para trafico y otro de 2 para reporting. Go no tiene pools de threads
//   que dimensionar — el runtime multiplexa goroutines sobre GOMAXPROCS hilos
//   del SO, y crear una goroutine cuesta ~2 KB. "Agotar el pool" no es un modo
//   de falla que exista aca.
//
//   Lo que SI existe es saturar el scheduler: una goroutine CPU-bound monopoliza
//   su procesador logico, y si hay tantas como GOMAXPROCS, las goroutines que
//   sirven trafico esperan. El sintoma final es el mismo (la operacion se
//   degrada); la causa raiz y el instrumento son distintos.
//
//   Por eso el aislamiento aca no es un pool: es un **semaforo de concurrencia**
//   (`chan struct{}` con capacidad N) que acota cuantos reportes corren a la vez,
//   dejando GOMAXPROCS-N procesadores libres para el trafico. Y el trabajo
//   pesado llama a `runtime.Gosched()` periodicamente para ceder el procesador
//   — el equivalente honesto del `Thread.yield()` de Java.
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"os"
	"runtime"
	"strconv"
	"sync/atomic"
	"time"
)

const (
	caseName          = "11 - Reportes pesados que bloquean operacion"
	reportingSlots    = 2
	degradedThreshold = 100 // ms
)

var stack = envOr("APP_STACK", "Go 1.23")

var (
	legacyReports       int64
	isolatedReports     int64
	orderWrites         int64
	orderWritesDegraded int64

	// inFlight cuenta requests en vuelo. Go no expone "threads activos" porque
	// no hay pool: esta es la medida equivalente y honesta.
	inFlight int64
	// reportingWaiting cuenta reportes esperando slot — el analogo de la cola.
	reportingWaiting int64
)

// reportingLimiter: semaforo de concurrencia. Sustituye al reportingPool de
// Java, pero acota trabajo simultaneo, no threads reservados.
var reportingLimiter = make(chan struct{}, reportingSlots)

// ---------- arranque ----------

func main() {
	mux := http.NewServeMux()
	mux.HandleFunc("/", route)

	port := envOr("PORT", "8080")
	log.Printf("[case11-go] listening on %s (GOMAXPROCS=%d)", port, runtime.GOMAXPROCS(0))
	if err := http.ListenAndServe(":"+port, mux); err != nil {
		log.Fatalf("listen: %v", err)
	}
}

// ---------- routing ----------

func route(w http.ResponseWriter, r *http.Request) {
	atomic.AddInt64(&inFlight, 1)
	defer atomic.AddInt64(&inFlight, -1)

	path := r.URL.Path
	rows := bounded(r.URL.Query().Get("rows"), 200000, 1000, 5000000)

	status := http.StatusOK
	var payload any

	switch path {
	case "/", "/index":
		payload = map[string]any{
			"case":  caseName,
			"stack": stack,
			"routes": []string{
				"/health", "/report-legacy?rows=200000", "/report-isolated?rows=200000",
				"/order-write", "/activity", "/diagnostics/summary", "/reset-lab",
			},
		}
	case "/health":
		payload = map[string]string{"status": "ok", "stack": stack, "case": caseName}
	case "/report-legacy":
		payload = reportLegacy(rows)
		atomic.AddInt64(&legacyReports, 1)
	case "/report-isolated":
		payload = reportIsolated(rows)
		atomic.AddInt64(&isolatedReports, 1)
	case "/order-write":
		payload = orderWrite()
		atomic.AddInt64(&orderWrites, 1)
	case "/activity":
		payload = activity()
	case "/diagnostics/summary":
		payload = diagnostics()
	case "/reset-lab":
		atomic.StoreInt64(&legacyReports, 0)
		atomic.StoreInt64(&isolatedReports, 0)
		atomic.StoreInt64(&orderWrites, 0)
		atomic.StoreInt64(&orderWritesDegraded, 0)
		payload = map[string]string{"status": "reset"}
	default:
		status = http.StatusNotFound
		payload = map[string]string{"error": "not_found", "path": path}
	}

	sendJSON(w, status, payload)
}

// ---------- trabajo pesado ----------

// crunch es el trabajo CPU-bound. Gosched() cede el procesador periodicamente:
// sin eso, una goroutine CPU-bound puede retener su P y demorar al resto.
func crunch(rows int, yield bool) int64 {
	var checksum int64
	for i := 0; i < rows; i++ {
		checksum += (int64(i) * 13) % 7
		if yield && (i&0xFFFF) == 0 {
			runtime.Gosched()
		}
	}
	return checksum
}

// reportLegacy: corre sin acotar en la goroutine del request. N reportes
// concurrentes ocupan N procesadores logicos y le quitan CPU al trafico.
func reportLegacy(rows int) map[string]any {
	start := time.Now()
	checksum := crunch(rows, true)
	return map[string]any{
		"variant": "legacy", "rows": rows, "checksum": checksum,
		"elapsed_ms":        time.Since(start).Milliseconds(),
		"ran_on_pool":       "request-goroutine (sin acotar)",
		"main_pool_active":  atomic.LoadInt64(&inFlight),
		"main_pool_queue":   atomic.LoadInt64(&reportingWaiting),
		"gomaxprocs":        runtime.GOMAXPROCS(0),
		"note":              "corre sin limite de concurrencia; mas reportes = menos CPU para /order-write.",
	}
}

// reportIsolated: adquiere un slot del limitador. Como maximo `reportingSlots`
// reportes corren a la vez, el resto espera — y GOMAXPROCS-N procesadores
// quedan disponibles para servir trafico.
func reportIsolated(rows int) map[string]any {
	start := time.Now()

	atomic.AddInt64(&reportingWaiting, 1)
	reportingLimiter <- struct{}{} // adquirir slot (bloquea si no hay)
	atomic.AddInt64(&reportingWaiting, -1)
	defer func() { <-reportingLimiter }() // liberar slot

	checksum := crunch(rows, true)
	return map[string]any{
		"variant": "isolated", "rows": rows, "checksum": checksum,
		"elapsed_ms":        time.Since(start).Milliseconds(),
		"ran_on_pool":       "reporting-limiter (max " + strconv.Itoa(reportingSlots) + " concurrentes)",
		"main_pool_active":  atomic.LoadInt64(&inFlight),
		"main_pool_queue":   atomic.LoadInt64(&reportingWaiting),
		"gomaxprocs":        runtime.GOMAXPROCS(0),
		"note":              "acotado por semaforo de concurrencia; /order-write conserva CPU disponible.",
	}
}

// orderWrite: escritura corta. Si tarda de mas, la operacion esta degradada.
func orderWrite() map[string]any {
	activeBefore := atomic.LoadInt64(&inFlight)
	start := time.Now()
	time.Sleep(20 * time.Millisecond)
	elapsedMs := time.Since(start).Milliseconds()

	degraded := elapsedMs > degradedThreshold
	if degraded {
		atomic.AddInt64(&orderWritesDegraded, 1)
	}
	note := "operacion responde con latencia normal"
	if degraded {
		note = "la latencia subio por saturacion de CPU del proceso"
	}
	return map[string]any{
		"variant": "order-write", "elapsed_ms": elapsedMs, "degraded": degraded,
		"main_pool_active_at_entry": activeBefore,
		"note":                      note,
	}
}

func activity() map[string]any {
	return map[string]any{
		"main_pool_active":      atomic.LoadInt64(&inFlight),
		"main_pool_queue":       atomic.LoadInt64(&reportingWaiting),
		"main_pool_size":        runtime.NumGoroutine(),
		"main_pool_max":         runtime.GOMAXPROCS(0),
		"reporting_slots":       reportingSlots,
		"reporting_slots_used":  len(reportingLimiter),
		"order_writes":          atomic.LoadInt64(&orderWrites),
		"order_writes_degraded": atomic.LoadInt64(&orderWritesDegraded),
	}
}

func diagnostics() map[string]any {
	return map[string]any{
		"stack": stack,
		"case":  caseName,
		"legacy": map[string]any{
			"reports":  atomic.LoadInt64(&legacyReports),
			"behavior": "reporte sin acotar en la goroutine del request; /order-write pierde CPU",
		},
		"isolated": map[string]any{
			"reports":  atomic.LoadInt64(&isolatedReports),
			"behavior": "semaforo de concurrencia acota reportes simultaneos; /order-write intacto",
		},
		"activity": activity(),
	}
}

// ---------- helpers ----------

func sendJSON(w http.ResponseWriter, status int, payload any) {
	body, err := json.Marshal(payload)
	if err != nil {
		http.Error(w, `{"error":"marshal"}`, http.StatusInternalServerError)
		return
	}
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(status)
	_, _ = w.Write(body)
}

func envOr(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

func bounded(raw string, fallback, min, max int) int {
	n, err := strconv.Atoi(raw)
	if err != nil {
		n = fallback
	}
	if n < min {
		return min
	}
	if n > max {
		return max
	}
	return n
}

File: cases/11-heavy-reporting-blocks-operations/rust/app/src/main.rs
// Caso 11 — Reportes pesados que bloquean operacion — stack Rust 1.83.
//
// Legacy: el reporte corre sin acotar en el thread del request.
// Isolated: el reporte pasa por un limitador de concurrencia; como maximo N
// reportes corren a la vez y siempre queda CPU para la operacion.
//
// El contraste que este stack aporta, y por que este caso NO se traduce literal:
//
//   Java y .NET aislan con pools de threads separados. Go no tiene pools, asi
//   que usa un semaforo de concurrencia. Rust esta en el mismo lugar que Go:
//   `std::thread::spawn` crea un thread del SO por conexion, sin pool que
//   dimensionar, y `std` no trae un ExecutorService que copiar.
//
//   Pero hay una diferencia importante con Go que este caso hace visible:
//   **un thread de Rust es un thread del SO** (~8 MB de stack virtual, ~2-8 KB
//   residentes), mientras una goroutine arranca en ~2 KB y el runtime la
//   multiplexa. Go puede tener cien mil goroutines; Rust con `std::thread` no
//   puede tener cien mil threads. El modelo thread-per-connection de este caso
//   es honesto para un lab, y seria la primera cosa a cambiar en produccion —
//   ahi se usa `tokio`, que multiplexa tareas como Go multiplexa goroutines.
//
//   El limitador se implementa con `Mutex<usize>` + `Condvar`: el thread que
//   no consigue slot espera a ser despertado en vez de hacer busy-wait. Es la
//   primitiva que `std` da para esto, y es mas explicita que el canal de Go.

use std::collections::HashMap;
use std::io::{BufRead, BufReader, Write};
use std::net::{TcpListener, TcpStream};
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::{Condvar, Mutex};
use std::thread;
use std::time::{Duration, Instant};

const CASE_NAME: &str = "11 - Reportes pesados que bloquean operacion";
const REPORTING_SLOTS: usize = 2;
const DEGRADED_THRESHOLD_MS: i64 = 100;

static LEGACY_REPORTS: AtomicI64 = AtomicI64::new(0);
static ISOLATED_REPORTS: AtomicI64 = AtomicI64::new(0);
static ORDER_WRITES: AtomicI64 = AtomicI64::new(0);
static ORDER_WRITES_DEGRADED: AtomicI64 = AtomicI64::new(0);
static IN_FLIGHT: AtomicI64 = AtomicI64::new(0);
static REPORTING_WAITING: AtomicI64 = AtomicI64::new(0);

/// Limitador de concurrencia: Mutex con el conteo de slots libres + Condvar
/// para dormir al que espera en vez de girar en vacio.
static SLOTS_USED: Mutex<usize> = Mutex::new(0);
static SLOT_FREED: Condvar = Condvar::new();

fn acquire_slot() {
    let mut used = SLOTS_USED.lock().unwrap();
    while *used >= REPORTING_SLOTS {
        used = SLOT_FREED.wait(used).unwrap();
    }
    *used += 1;
}

fn release_slot() {
    let mut used = SLOTS_USED.lock().unwrap();
    *used = used.saturating_sub(1);
    SLOT_FREED.notify_one();
}

fn slots_used() -> usize {
    *SLOTS_USED.lock().unwrap()
}

fn stack() -> String {
    std::env::var("APP_STACK").unwrap_or_else(|_| "Rust 1.83".to_string())
}

fn main() {
    let port: u16 = std::env::var("PORT").ok().and_then(|p| p.parse().ok()).unwrap_or(8080);
    let listener = TcpListener::bind(("0.0.0.0", port)).expect("bind");
    println!(
        "[case11-rust] listening on {port} (cpus={})",
        thread::available_parallelism().map(|n| n.get()).unwrap_or(1)
    );
    for stream in listener.incoming().flatten() {
        thread::spawn(move || handle_conn(stream));
    }
}

// ---------- capa HTTP minima ----------

fn handle_conn(mut stream: TcpStream) {
    IN_FLIGHT.fetch_add(1, Ordering::Relaxed);
    let mut reader = BufReader::new(match stream.try_clone() {
        Ok(s) => s,
        Err(_) => {
            IN_FLIGHT.fetch_sub(1, Ordering::Relaxed);
            return;
        }
    });
    let mut request_line = String::new();
    if reader.read_line(&mut request_line).is_err() {
        IN_FLIGHT.fetch_sub(1, Ordering::Relaxed);
        return;
    }
    loop {
        let mut line = String::new();
        match reader.read_line(&mut line) {
            Ok(0) => break,
            Ok(_) if line.trim().is_empty() => break,
            Ok(_) => continue,
            Err(_) => break,
        }
    }
    let target = request_line.split_whitespace().nth(1).unwrap_or("/");
    let (path, query) = target.split_once('?').unwrap_or((target, ""));
    let params = parse_query(query);
    let (status, body) = route(path, &params);
    let reason = if status == 200 { "OK" } else { "Not Found" };
    let response = format!(
        "HTTP/1.1 {status} {reason}\r\nContent-Type: application/json; charset=utf-8\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{body}",
        body.as_bytes().len()
    );
    let _ = stream.write_all(response.as_bytes());
    let _ = stream.flush();
    IN_FLIGHT.fetch_sub(1, Ordering::Relaxed);
}

fn parse_query(raw: &str) -> HashMap<String, String> {
    let mut out = HashMap::new();
    for pair in raw.split('&').filter(|p| !p.is_empty()) {
        let (k, v) = pair.split_once('=').unwrap_or((pair, ""));
        out.insert(k.to_string(), v.to_string());
    }
    out
}

// ---------- routing ----------

fn route(path: &str, params: &HashMap<String, String>) -> (u16, String) {
    let rows = bounded(params.get("rows").map(String::as_str), 200000, 1000, 5000000);

    match path {
        "/" | "/index" => (
            200,
            format!(
                r#"{{"case":"{CASE_NAME}","stack":"{}","routes":["/health","/report-legacy?rows=200000","/report-isolated?rows=200000","/order-write","/activity","/diagnostics/summary","/reset-lab"]}}"#,
                stack()
            ),
        ),
        "/health" => (
            200,
            format!(r#"{{"status":"ok","stack":"{}","case":"{CASE_NAME}"}}"#, stack()),
        ),
        "/report-legacy" => {
            let body = report_legacy(rows);
            LEGACY_REPORTS.fetch_add(1, Ordering::Relaxed);
            (200, body)
        }
        "/report-isolated" => {
            let body = report_isolated(rows);
            ISOLATED_REPORTS.fetch_add(1, Ordering::Relaxed);
            (200, body)
        }
        "/order-write" => {
            let body = order_write();
            ORDER_WRITES.fetch_add(1, Ordering::Relaxed);
            (200, body)
        }
        "/activity" => (200, activity()),
        "/diagnostics/summary" => (200, diagnostics()),
        "/reset-lab" => {
            LEGACY_REPORTS.store(0, Ordering::Relaxed);
            ISOLATED_REPORTS.store(0, Ordering::Relaxed);
            ORDER_WRITES.store(0, Ordering::Relaxed);
            ORDER_WRITES_DEGRADED.store(0, Ordering::Relaxed);
            (200, r#"{"status":"reset"}"#.to_string())
        }
        _ => (
            404,
            format!(r#"{{"error":"not_found","path":"{}"}}"#, escape(path)),
        ),
    }
}

// ---------- trabajo pesado ----------

fn crunch(rows: i64) -> i64 {
    let mut checksum: i64 = 0;
    for i in 0..rows {
        checksum += (i * 13) % 7;
        if (i & 0xFFFF) == 0 {
            thread::yield_now();
        }
    }
    checksum
}

/// Legacy: corre sin acotar en el thread del request.
fn report_legacy(rows: i64) -> String {
    let start = Instant::now();
    let checksum = crunch(rows);
    format!(
        r#"{{"variant":"legacy","rows":{rows},"checksum":{checksum},"elapsed_ms":{},"ran_on_pool":"request-thread (sin acotar)","main_pool_active":{},"main_pool_queue":{},"cpus":{},"note":"corre sin limite de concurrencia; mas reportes = menos CPU para /order-write."}}"#,
        start.elapsed().as_millis(),
        IN_FLIGHT.load(Ordering::Relaxed),
        REPORTING_WAITING.load(Ordering::Relaxed),
        cpus()
    )
}

/// Isolated: adquiere un slot; el que no consigue duerme en la Condvar.
fn report_isolated(rows: i64) -> String {
    let start = Instant::now();
    REPORTING_WAITING.fetch_add(1, Ordering::Relaxed);
    acquire_slot();
    REPORTING_WAITING.fetch_sub(1, Ordering::Relaxed);

    let checksum = crunch(rows);
    release_slot();

    format!(
        r#"{{"variant":"isolated","rows":{rows},"checksum":{checksum},"elapsed_ms":{},"ran_on_pool":"reporting-limiter (max {REPORTING_SLOTS} concurrentes)","main_pool_active":{},"main_pool_queue":{},"cpus":{},"note":"acotado por Mutex+Condvar; /order-write conserva CPU disponible."}}"#,
        start.elapsed().as_millis(),
        IN_FLIGHT.load(Ordering::Relaxed),
        REPORTING_WAITING.load(Ordering::Relaxed),
        cpus()
    )
}

fn order_write() -> String {
    let active_before = IN_FLIGHT.load(Ordering::Relaxed);
    let start = Instant::now();
    thread::sleep(Duration::from_millis(20));
    let elapsed_ms = start.elapsed().as_millis() as i64;

    let degraded = elapsed_ms > DEGRADED_THRESHOLD_MS;
    if degraded {
        ORDER_WRITES_DEGRADED.fetch_add(1, Ordering::Relaxed);
    }
    let note = if degraded {
        "la latencia subio por saturacion de CPU del proceso"
    } else {
        "operacion responde con latencia normal"
    };
    format!(
        r#"{{"variant":"order-write","elapsed_ms":{elapsed_ms},"degraded":{degraded},"main_pool_active_at_entry":{active_before},"note":"{note}"}}"#
    )
}

fn activity() -> String {
    format!(
        r#"{{"main_pool_active":{},"main_pool_queue":{},"main_pool_max":{},"reporting_slots":{REPORTING_SLOTS},"reporting_slots_used":{},"order_writes":{},"order_writes_degraded":{}}}"#,
        IN_FLIGHT.load(Ordering::Relaxed),
        REPORTING_WAITING.load(Ordering::Relaxed),
        cpus(),
        slots_used(),
        ORDER_WRITES.load(Ordering::Relaxed),
        ORDER_WRITES_DEGRADED.load(Ordering::Relaxed)
    )
}

fn diagnostics() -> String {
    format!(
        r#"{{"stack":"{}","case":"{CASE_NAME}","legacy":{{"reports":{},"behavior":"reporte sin acotar en el thread del request; /order-write pierde CPU"}},"isolated":{{"reports":{},"behavior":"Mutex+Condvar acota reportes simultaneos; /order-write intacto"}},"activity":{}}}"#,
        stack(),
        LEGACY_REPORTS.load(Ordering::Relaxed),
        ISOLATED_REPORTS.load(Ordering::Relaxed),
        activity()
    )
}

// ---------- helpers ----------

fn cpus() -> usize {
    thread::available_parallelism().map(|n| n.get()).unwrap_or(1)
}

fn bounded(raw: Option<&str>, dflt: i64, min: i64, max: i64) -> i64 {
    raw.and_then(|r| r.parse::<i64>().ok()).unwrap_or(dflt).clamp(min, max)
}

fn escape(v: &str) -> String {
    v.replace('\\', "\\\\").replace('"', "\\\"")
}
